File: src/workspaces.py
#!/usr/bin/env python3
"""Workspace registry for agent-pipeline.

Each requirement can target a named local workspace. The name is safe to expose
in Feishu Base, while absolute paths stay in local config.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import config as C

_WORKSPACE_TOKEN_RE = re.compile(r"(?:^|\s)#(?P<key>[A-Za-z0-9_.-]+)(?=\s|[：:]|$)")


@dataclass(frozen=True)
class Workspace:
    key: str
    path: Path
    base_ref: str = "origin/main"
    scm: str = "git"
    test_cmd: str = ""
    code_exts: tuple[str, ...] = ()
    push_enabled: bool = False
    pr_enabled: bool = False
    pr_provider: str = "github"   # github | gitlab | none（svn 忽略此项）
    gh_repo: str = ""             # 可选：PR 目标仓库 org/repo（github）

    @property
    def safe_key(self) -> str:
        return re.sub(r"[^A-Za-z0-9_.-]+", "_", self.key).strip("._-") or "default"


def _load_raw() -> dict[str, Any]:
    if not C.WORKSPACES_FILE.exists():
        return {}
    try:
        return json.loads(C.WORKSPACES_FILE.read_text(encoding="utf-8"))
    except Exception as exc:
        raise RuntimeError(f"读取 workspaces.json 失败: {exc}") from exc


def _default_key(raw: dict[str, Any]) -> str:
    value = raw.get("default")
    if isinstance(value, str) and value.strip():
        return value.strip()
    items = raw.get("items")
    if isinstance(items, dict) and items:
        return next(iter(items))
    return "default"


def _fallback_workspace(key: str = "default") -> Workspace:
    return Workspace(
        key=key,
        path=C.REPO_PATH,
        base_ref=C.BASE_REF,
        test_cmd=C.TEST_CMD,
        code_exts=C.CODE_EXTS,
        push_enabled=C.PUSH_ENABLED,
        pr_enabled=C.PR_ENABLED,
    )


def get(key: str | None = None) -> Workspace:
    raw = _load_raw()
    items = raw.get("items") if isinstance(raw.get("items"), dict) else {}
    selected = (key or "").strip() or _default_key(raw)
    if not items:
        return _fallback_workspace(selected)
    cfg = items.get(selected)
    if cfg is None:
        allowed = ", ".join(sorted(items))
        raise ValueError(f"未知工作区 `{selected}`，可选：{allowed}")
    if isinstance(cfg, str):
        cfg = {"path": cfg}
    if not isinstance(cfg, dict):
        raise ValueError(f"工作区 `{selected}` 配置必须是对象或路径字符串")
    path = Path(str(cfg.get("path") or "")).expanduser()
    if not path:
        raise ValueError(f"工作区 `{selected}` 缺少 path")
    code_exts_raw = cfg.get("code_exts")
    if isinstance(code_exts_raw, str):
        code_exts = tuple(x.strip() for x in code_exts_raw.split(",") if x.strip())
    elif isinstance(code_exts_raw, list):
        code_exts = tuple(str(x).strip() for x in code_exts_raw if str(x).strip())
    else:
        code_exts = C.CODE_EXTS
    return Workspace(
        key=selected,
        path=path,
        base_ref=str(cfg.get("base") or C.BASE_REF),
        scm=str(cfg.get("scm") or "git"),
        test_cmd=str(cfg.get("test_cmd") if cfg.get("test_cmd") is not None else C.TEST_CMD),
        code_exts=code_exts,
        push_enabled=bool(cfg.get("push_enabled", C.PUSH_ENABLED)),
        pr_enabled=bool(cfg.get("pr_enabled", C.PR_ENABLED)),
        pr_provider=str(cfg.get("pr_provider") or "github").lower(),
        gh_repo=str(cfg.get("gh_repo") or C.GH_REPO),
    )


def list_workspaces() -> list[Workspace]:
    raw = _load_raw()
    items = raw.get("items") if isinstance(raw.get("items"), dict) else {}
    if not items:
        return [get(None)]
    return [get(key) for key in sorted(items)]


def parse_workspace_token(text: str) -> tuple[str, str | None]:
    """Return (text_without_workspace_token, workspace_key)."""
    match = _WORKSPACE_TOKEN_RE.search(text)
    if not match:
        return text, None
    key = match.group("key")
    cleaned = (text[: match.start()] + " " + text[match.end() :]).strip()
    cleaned = re.sub(r"\s+", " ", cleaned)
    return cleaned, key


def ensure_example() -> None:
    if C.WORKSPACES_FILE.exists():
        return
    default_name = C.REPO_PATH.name or "default"
    data = {
        "default": default_name,
        "items": {
            default_name: {
                "path": str(C.REPO_PATH),
                "scm": "git",
                "base": C.BASE_REF,
                "test_cmd": C.TEST_CMD,
            }
        },
    }
    C.WORKSPACES_FILE.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

File: go/workspaces.go
package main

import (
	"encoding/json"
	"os"
	"regexp"
	"strings"
	"sync"
)

type Workspace struct {
	Key          string
	Path         string
	SCM          string // git | svn
	WorkMode     string // worktree | inline
	BaseRef      string
	TargetBranch string
	TestCmd      string
	PRProvider   string // none | github | gitlab
	GHRepo       string
	PushEnabled  bool
	PREnabled    bool
}

type wsFile struct {
	Default string            `json:"default"`
	Items   map[string]wsItem `json:"items"`
}
type wsItem struct {
	Path         string `json:"path"`
	SCM          string `json:"scm"`
	WorkMode     string `json:"work_mode"`
	Base         string `json:"base"`
	TargetBranch string `json:"target_branch"`
	TestCmd      string `json:"test_cmd"`
	PRProvider   string `json:"pr_provider"`
	GHRepo       string `json:"gh_repo"`
	PushEnabled  bool   `json:"push_enabled"`
	PREnabled    bool   `json:"pr_enabled"`
}

var (
	wsCache *wsFile
	wsOnce  sync.Once
)

// loadWorkspaces 只完整初始化一次（sync.Once）：避免并发首调时读到半填充的缓存，
// 否则不同 goroutine 解析出的工作区/锁 key 会不一致，按工作区的串行锁形同虚设。
func loadWorkspaces() *wsFile {
	wsOnce.Do(func() {
		c := &wsFile{Items: map[string]wsItem{}}
		if b, err := os.ReadFile(cfg.WorkspacesFile); err == nil {
			json.Unmarshal(b, c)
		}
		wsCache = c
	})
	return wsCache
}

// defaultWorkspace 从 .env 合成（没有 workspaces.json 时）。
func defaultWorkspace() Workspace {
	return Workspace{
		Key: "default", Path: cfg.RepoPath, SCM: "git", WorkMode: "inline", BaseRef: cfg.BaseRef,
		TargetBranch: lastSeg(cfg.BaseRef), TestCmd: cfg.TestCmd, PRProvider: "none",
		GHRepo: cfg.GHRepo, PushEnabled: cfg.PushEnabled, PREnabled: cfg.PREnabled,
	}
}

func workspaceGet(key string) (Workspace, error) {
	f := loadWorkspaces()
	if len(f.Items) == 0 {
		if key == "" || key == "default" {
			return defaultWorkspace(), nil
		}
		return Workspace{}, errf("未知工作区 `%s`：未配置 workspaces.json", key)
	}
	if key == "" {
		key = f.Default
	}
	it, ok := f.Items[key]
	if !ok {
		var keys []string
		for k := range f.Items {
			keys = append(keys, k)
		}
		return Workspace{}, errf("未知工作区 `%s`，可选：%s", key, strings.Join(keys, ", "))
	}
	scm := it.SCM
	if scm == "" {
		scm = "git"
	}
	mode := it.WorkMode
	if mode == "" {
		mode = "inline"
	}
	base := it.Base
	if base == "" {
		base = "origin/main"
	}
	tb := it.TargetBranch
	if tb == "" {
		tb = lastSeg(base)
	}
	prov := it.PRProvider
	if prov == "" {
		prov = "none"
	}
	return Workspace{
		Key: key, Path: it.Path, SCM: scm, WorkMode: mode, BaseRef: base, TargetBranch: tb,
		TestCmd: it.TestCmd, PRProvider: prov, GHRepo: it.GHRepo,
		PushEnabled: it.PushEnabled, PREnabled: it.PREnabled,
	}, nil
}

func workspaceKeys() []string {
	f := loadWorkspaces()
	if len(f.Items) == 0 {
		return []string{"default"}
	}
	var keys []string
	for k := range f.Items {
		keys = append(keys, k)
	}
	return keys
}

func (w Workspace) safeKey() string {
	return regexp.MustCompile(`[^A-Za-z0-9_.-]`).ReplaceAllString(w.Key, "_")
}

var wsTokenRe = regexp.MustCompile(`#([A-Za-z0-9_.-]+)`)

// parseWorkspaceToken 从文本里抽 #key（返回去掉 token 的文本 + key）。
func parseWorkspaceToken(body string) (string, string) {
	m := wsTokenRe.FindStringSubmatch(body)
	if m == nil {
		return body, ""
	}
	cleaned := strings.TrimSpace(wsTokenRe.ReplaceAllString(body, ""))
	return cleaned, m[1]
}

func lastSeg(ref string) string {
	if i := strings.LastIndexByte(ref, '/'); i >= 0 {
		return ref[i+1:]
	}
	return ref
}
